---
title: "Running univariate and bivariate LAVA"
author: 
- name: "Aaron Wagen & Regina H. Reynolds"
  affiliation: UCL
output: 
  bookdown::html_document2:
    figure_caption: yes
    code_folding: hide
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include = FALSE}

library(circlize) # For chord diagram
library(ggplot2) # For plotting
library(ggraph) # For edge diagrams
library(janitor) # For cleaning dirty data
library(tidyverse) # For tidy manipulation of data
library(stringr) # For string manipulation
library(rtracklayer) # For loading of reference gtf

knitr::opts_chunk$set(echo = T, warning = F, message = F)

# Set defaults for ggplots 
theme_rhr <- 
  theme_bw(base_family = "Helvetica",
           base_size = 10) + 
  theme(panel.grid.major.x = element_blank(),
        legend.position = "top",
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        axis.title.y = element_text(vjust = 0.6),
        panel.spacing = unit(0.1, "lines"))

loci <- LAVA::read.loci(here::here("results", "01_input_prep", "gwas_filtered.loci"))
fct_disease <- c("AD", "LBD", "PD", "BIP", "MDD", "SCZ")

```

> Aim: run univariate and bivariate tests
<br><br>

# Background

Neurodegenerative diseases are often defined by the predominant neuropathology (e.g. synucleinopathies by $\alpha$-synuclein deposition), but more often than not, they also present with co-pathologies, suggesting that they might share common pathogenic pathways (PMID: [29878075](https://pubmed.ncbi.nlm.nih.gov/29878075/), [30258235](https://pubmed.ncbi.nlm.nih.gov/30258235/)). This notion is supported by genetic studies, which have (i) identified shared risk loci across neurodegenerative disorders, such as APOE and BIN1 in Alzheimer’s disease (AD) and Lewy body dementia (LBD), or GBA, SNCA, TMEM175 in Parkinson’s disease (PD) and LBD and (ii) demonstrated that genetic risk scores derived from one neurodegenerative disorder can predict risk of another, as with AD and PD predicting risk of LBD (PMID: [31701892](https://pubmed.ncbi.nlm.nih.gov/31701892/), [30953760](https://pubmed.ncbi.nlm.nih.gov/30953760/), [33589841](https://pubmed.ncbi.nlm.nih.gov/33589841/)).

From a clinical perspective, neurodegenerative diseases are often also defined in terms of their predominant cognitive and/or motor symptoms (e.g. Alzheimer's disease by cognitive and memory impairment or Parkinson's disease by parkinsonism), but in reality, present as highly heterogenous diseases, with symptoms spanning multiple domains, including neurospychiatric symptoms (PMID: [27188934](https://pubmed.ncbi.nlm.nih.gov/27188934/), [28332488](https://pubmed.ncbi.nlm.nih.gov/28332488/). Indeed, a higher prevalence of depression has been observed in individuals with dementia compared to those without dementia (PMID: [30536144](https://pubmed.ncbi.nlm.nih.gov/30536144/)). Notably, a similar (albeit reversed) phenomenon has been observed in some neuropsychiatric disorders, with a higher risk of dementia diagnoses observed in individuals with schizophrenia (SCZ) versus individuals without a history of serious mental illness (PMID: [33688938](https://pubmed.ncbi.nlm.nih.gov/33688938/), [26444987](https://pubmed.ncbi.nlm.nih.gov/26444987/)).

These clinical and/or pathological overlaps are not, however, always reflected in terms of global genetic correlations, with a recent study of global genetic correlation between neurological phenotypes demonstrating limited overlap between individual neurodegenerative disorders as well as between neurodegenerative and neuropsychiatric disorders (PMID: [29930110](https://pubmed.ncbi.nlm.nih.gov/29930110/)). One explanation for this is that global studies only consider an average across the entire genome, and thus, may fail to detect correlations confined to particular genomic regions, or where changes are balanced out over the genome. This limitation has recently been addressed with the development of Local analysis of [co]variant annotation (LAVA), which is able to evaluate local heritability over multiple traits of interest (using GWAS summary statistics) and detect local regions of shared genetic association (https://www.biorxiv.org/content/10.1101/2020.12.31.424652v2). Here, we apply LAVA to genome-wide association studies (GWASs) derived from three neurodenerative disorders (AD, PMID: [30617256](https://pubmed.ncbi.nlm.nih.gov/30617256/);LBD, PMID: [33589841](https://pubmed.ncbi.nlm.nih.gov/33589841/); and PD, PMID: [31701892](https://pubmed.ncbi.nlm.nih.gov/31701892/)), and three neuropsychiatric disorders (bipolar disorder, BIP, PMID: [34002096](https://pubmed.ncbi.nlm.nih.gov/34002096/); major depressive disorder, MDD, PMID: [30718901](https://pubmed.ncbi.nlm.nih.gov/30718901/); and SCZ, PMID: [29483656](https://pubmed.ncbi.nlm.nih.gov/29483656/)).

# Supplementary code {.tabset}

Following section includes any intermediary code used in this `.Rmd`.

## Run univariate and bivariate tests
This can either be sourced directly:

```{r run-univar-bivar, eval = F}

source(here::here("scripts", "02_run_univar_bivar_test.R"))

```

Alternatively, for many phenotypes/loci it can be worth running this script directly on the cluster, using `nohup`, as performed below:

```{bash bash-run-univar-bivar, eval = F}
# Have to navigate to root project folder for script to work (as it uses here package)
cd /home/rreynolds/misc_projects/neurodegen-psych-local-corr

nohup Rscript \
/home/rreynolds/misc_projects/neurodegen-psych-local-corr/scripts/02_run_univar_bivar_test.R \
&>/home/rreynolds/misc_projects/neurodegen-psych-local-corr/logs/02_run_univar_bivar_test.log&

```

## Tidying and loading results
Once run, results can be loaded using the following code chunk:

```{r tidy-lava-results, eval = F}

source(here::here("scripts", "02b_tidy_univar_bivar.R"))

```

```{r load-lava-results}

results <- readRDS(here::here("results", "02_univar_bivar_test", "results_summary.rds"))

```

```{r load-ldsc}
global <- 
  read_delim(
    file = here::here("results", "01_input_prep", "ldsc_corr", "ldsc_correlations.txt"),
    delim = "\t"
    ) 

```

## Filtering for pleiotropic genes
```{r pleiotropic-genes}

ref <- import("/data/references/ensembl/gtf_gff3/v87/Homo_sapiens.GRCh37.87.gtf")
ref <- ref %>% keepSeqlevels(c(1:22), pruning.mode = "coarse") 
ref <- ref[ref$type == "gene"]

# genes to check
genes <- 
  c("APOE", "TMEM175", "SNCA", "BIN1", "GRN", "KIF5A")

# construct granges object with LD blocks for overlap with reference gtf
loci_gr <-
  loci %>%
  GenomicRanges::makeGRangesFromDataFrame(
    .,
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    seqinfo = NULL,
    seqnames.field = "CHR",
    start.field = "START",
    end.field = "STOP"
  )

overlap <-
  GenomicRanges::findOverlaps(loci_gr, ref) %>%
  tibble::as_tibble()

loci_genes <-  
  tibble::tibble(
    locus = loci_gr[overlap$queryHits]$LOC,
    chr = loci_gr[overlap$queryHits] %>% GenomeInfoDb::seqnames() %>% as.character(),
    locus_start = loci_gr[overlap$queryHits] %>% BiocGenerics::start(),
    locus_end = loci_gr[overlap$queryHits] %>% BiocGenerics::end(),
    gene_id = ref[overlap$subjectHits]$gene_id,
    gene_name =
      ref[overlap$subjectHits]$gene_name %>%
      stringr::str_replace_all("-", "_") %>%
      stringr::str_replace_all("\\.", "_"),
    gene_biotype = ref[overlap$subjectHits]$gene_biotype %>% as.factor(),
    gene_strand = ref[overlap$subjectHits] %>% BiocGenerics::strand() %>% as.character(),
    gene_start = ref[overlap$subjectHits] %>% BiocGenerics::start(),
    gene_end = ref[overlap$subjectHits] %>% BiocGenerics::end()
  )

# filter df with ld block & gene info for genes of interest
loci_genes_filtered <- 
  loci_genes %>% 
  dplyr::filter(
    gene_name %in% genes
  )

```

## Loading plotting functions

```{r load-plot-func}

source(here::here("R", "plot_global_corr.R"))
source(here::here("R", "plot_chord_diagram.R"))
source(here::here("R", "plot_locus.R"))
source(here::here("R", "plot_edge_diagram.R"))

```

# Methods {.tabset}

## Running univariate and bivariate tests
The detection of valid and interpretable local genetic corrlation ($r_{g}$) requires the presence of sufficient local genetic signal. For this reason, a univariate test was performed as a filtering step for bivariate local $r_{g}$ analyses. Bivariate local correlations were only performed for pairs of traits which both exhibited a significant univariate local genetic signal (p < 0.05/`r nrow(loci)`, where the denominator represents the total number of tested loci, i.e. LD blocks). This resulted in a total of **`r nrow(results$bivar)`** bivariate tests spanning **`r results$bivar %>% .[["locus"]] %>% unique() %>% length()`** distinct LD blocks. 
<br><br>

# Results {.tabset}

## Tabular results {.tabset}

### Global correlations
```{r global-corr-table}

print("Global correlation column descriptions:")

tibble(
  column = colnames(global),
  description = 
    c(
      "Phenotype 1",
      "Phenotype 2",
      "The estimated genetic correlation",
      "The bootstrap standard error of the genetic correlation estimate",
      "The bootstrap standard error of the genetic correlation estimate",
      "P-value for genetic correlation",
      "Estimated snp-heritability of the second phenotype ",
      "Standard error of h2 for phenotype 2",
      "Single-trait LD score regression intercept for phenotype 2",
      "Standard error for single-trait LD score regression intercept for phenotype 2",
      "Estimated genetic covariance between p1 and p2",
      "Bootstrap standard error of cross-trait LD score regression intercept"
    )
) %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')


print("Global correlation results (p < 0.05/n_tests):")

combn <-
  global$p1 %>%
  unique() %>%
  combn(m = 2) %>%
  t() %>%
  as_tibble() %>%
  dplyr::rename(
    p1 = V1,
    p2 = V2
  )

global_signif <- 
  global %>% 
  dplyr::filter(
    !p1==p2
  ) %>% 
  dplyr::inner_join(combn) %>%
  dplyr::filter(
    p < 0.05/nrow(combn)
  )

global_signif %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

```


### Univariate results
```{r univar-table}
print("Univariate column descriptions:")

tibble(
  column = colnames(results$univ),
  description = 
    c(
      "LD block ID",
      "LD block chromosome",
      "LD block start base pair",
      "LD block end base pair",
      "The number of SNPs within the LD block",
      "The number of PCs retained within the LD block",
      "Analysed phenotype",
      "Observed local heritability",
      "P-value from the univariate test (F-test for continuous, Chi-sq for binary)"
    )
) %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

print("Univariate results (p < 0.05/n_ld_blocks):")

results$univ %>% 
  dplyr::filter(p < 0.05/nrow(loci)) %>%   
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')
```

### Bivariate results
```{r bivar-table}

print("Bivariate column descriptions:")

tibble(
  column = colnames(results$bivar),
  description = 
    c(
      "LD block ID",
      "LD block chromosome",
      "LD block start base pair",
      "LD block end base pair",
      "The number of SNPs within the LD block",
      "The number of PCs retained within the LD block",
      "Phenotype 1",
      "Phenotype 2",
      "Standardised coefficient for the local genetic correlation",
      "Lower 95% confidence estimate for rho",
      "Upper 95% confidence estimate for rho",
      "Equivalent of taking the square of rho. Denotes the proportion of variance in genetic signal for phen1 explained by phen2 (and vice versa)",
      "Lower 95% confidence estimate for r2",
      "Upper 95% confidence estimate for r2",
      "Simulation p-values for the local genetic correlation",
      "FDR-corrected p-value"
    )
) %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

print("Bivariate results (p < 0.05/n_bivar_tests):")

bivar_thres <- 0.05/nrow(results$bivar)

bivar_bonf <- 
  results$bivar %>% 
  dplyr::filter(p < bivar_thres)

bivar_bonf %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')


```

## Global correlations versus bivariate tests {.tabset}

### Text
- Global cross-trait correlations revealed `r nrow(global_signif)` significant global $r_{g}$'s (Figure \@ref(fig:global-corr)). These global $r_{g}$'s are very much in line with what we expect from previous literature e.g. significant positive correlations between the three neuropsychiatric disorders (PMID: [29930110](https://pubmed.ncbi.nlm.nih.gov/29930110/)). Between neurodegenerative disorders, we only observe a significant postive $r_{g}$ between LBD and PD.
- With a Bonferroni cut-off of `r signif(bivar_thres, digits = 2) %>% format(scientific = F)`, we detect a total of `r bivar_bonf %>% nrow()` significant bivariate local $r_{g}$'s across `r bivar_bonf %>% dplyr::distinct(locus) %>% nrow()` distinct LD blocks, of which `r bivar_bonf %>% dplyr::count(locus) %>% dplyr::filter(n > 1) %>% nrow()` were associated with more than one trait pair. Notably, we detected local genetic correlations between pairs of phenotypes in the absence of a significant global genetic correlation (\@ref(fig:global-local-plot)).
- The phenotype with the highest number of significant bivariate local $r_{g}$'s was SCZ followed by BIP. Notably, the number of significant bivariate local $r_{g}$'s did not appear to correlate with GWAS sample size (Figure \@ref(fig:bivar-sample-size)). 
- As seen from the table below and Figure \@ref(fig:chord-diagram-rho), the phenotype pairs with the most local $r_{g}$'s were (i) BIP and SCZ and (ii) AD and PD. 
- For `r bivar_bonf %>% dplyr::filter(r2.upper == 1) %>% nrow()` local $r_{g}$'s, the 95% confidence intervals (CI’s) for the explained variance (r2) included 1 (Figure \@ref(fig:bivar-correlations-bonf-phen-count)), which is consistent with a scenario where the genetic signal of those pairs of phenotypes in these LD blocks is completely shared.  
- Within neuropsychiatric disorders, all significant bivariate local $r_{g}$'s were positive (matching the overall positive global $r_{g}$), while within neurodegenerative disorders there was a mix of positive and negative local $r_{g}$'s (Figure \@ref(fig:chord-diagram-rho)).
- All three neurodegenerative disorders shared one significant local $r_{g}$ with SCZ, and also PD shared one negative local $r_{g}$ with BIP. 

### Figures

```{r gwas-n, fig.cap = "Heatmap of samples numbers in each GWAS."}

data_to_plot <- 
  read_delim(
    file = file.path(here::here("results", "01_input_prep"),
                     "input.info.txt"),
    delim = "\t"
  ) %>% 
  dplyr::select(-filename) %>% 
  tidyr::pivot_longer(
    cols = cases:controls,
    names_to = "group",
    values_to = "total"
  ) %>% 
  dplyr::mutate(
     phenotype_group =
      case_when(
        phenotype %in% c("AD2019", "LBD2020", "PD2019.meta5.ex23andMe") ~ "Neurodegenerative",
        TRUE ~ "Neuropsychiatric"
      ),
    phenotype =
      case_when(
        phenotype == "AD2019" ~ "Alzheimer's disease\n(AD)",
        phenotype == "LBD2020" ~ "Lewy body dementia\n(LBD)",
        phenotype == "PD2019.meta5.ex23andMe" ~ "Parkinson's disease\n(PD)",
        phenotype == "BIP2021" ~ "Bipolar disorder (BIP)",
        phenotype == "MDD2019" ~ "Major depressive disorder\n(MDD)",
        phenotype == "SCZ2018" ~ "Schizophrenia\n(SCZ)"
      )
  )

plot <- 
  data_to_plot %>% 
  dplyr::mutate(
    n_text_col = 
      case_when(
        total > 200009 ~ "high",
        TRUE ~ "low"
      ),
    group = stringr::str_to_title(group)
  ) %>% 
  ggplot(
    aes(
      x = group,
      y = 
        fct_relevel(
          phenotype,
          c(data_to_plot %>% dplyr::arrange(phenotype_group, phenotype) %>% dplyr::pull(phenotype) %>% unique() %>% rev())
        ),
      fill = total
    )
  ) +
  geom_tile(colour = "black") +
  geom_text(
    aes(
      label = total %>% scales::comma(),
      colour = n_text_col
    ), 
    size = 3
  ) +
  coord_equal() +
  scale_x_discrete(
    position = "top"
    ) +
  labs(x = "", y = "") +
  scale_fill_distiller(palette = "Greys", direction = 1) +
  scale_colour_manual(values = c("white", "black")) +
  guides(colour = F, fill = F) +
  theme_minimal() + 
  theme(
    panel.grid = element_blank(),
  )

plot
```

```{r adpd-plot, eval = F, echo = F}

ggsave(
  plot,
  device = "png",
  path = here::here("tmp"), 
  filename = "20220207_gwas_n.png",
  height = 120,
  width = 100,
  units = "mm",
  dpi = 300,
  limitsize = TRUE
)

```

```{r global-corr, fig.cap = "Global genetic correlations between pairs of phenotypes. Significant negative and positive correlations are indicated by blue and red fill, respectively. Non-significant correlations (p >= 0.05/n_tests) have a grey fill."}

plot_global_corr(
  global_corr = global,
  n_phenotypes = 6
)

```

```{r global-local-plot, fig.cap = "Comparison between the global genetic correlation estimated by LDSC (bottom) and the mean local genetic correlation from LAVA (top) across all tested LD blocks. Significant global correlations (p < 0.05/n_tests) are indicated with an asterisk ('*'). The number of significant local genetic correlations (p < 0.05/n_tests) is indicated by a number in each tile."}

plot_global_bivar(
  global_corr = 
    global %>%
    dplyr::mutate(
      p1 = p1 %>%
        stringr::str_replace_all("[:digit:]", "") %>%
        stringr::str_remove("\\..*"),
      p2 = p2 %>%
        stringr::str_replace_all("[:digit:]", "") %>%
        stringr::str_remove("\\..*")
    ),
  bivar_corr = 
    results$bivar %>%
    dplyr::mutate(
      phen1 = phen1 %>%
        stringr::str_replace_all("[:digit:]", "") %>%
        stringr::str_remove("\\..*"),
      phen2 = phen2 %>%
        stringr::str_replace_all("[:digit:]", "") %>%
        stringr::str_remove("\\..*")
    ),
  n_phenotypes = 6
)

```


```{r bivar-sample-size, fig.cap = "Plot of total number of significant bivariate local genetic correlations for each phenotype across all LD blocks versus GWAS sample size."}

bivar_bonf %>% 
  dplyr::count(phen1) %>% 
  dplyr::bind_rows(
    bivar_bonf %>% 
      dplyr::count(phen2) %>% 
      dplyr::rename(phen1 = phen2)
    ) %>% 
  dplyr::group_by(phen1) %>% 
  dplyr::summarise(
    n_bivar = sum(n)
  ) %>% 
  dplyr::inner_join(
    read_delim(
      file = file.path(here::here("results", "01_input_prep"),
                       "input.info.txt"),
      delim = "\t"
    ) %>% 
    dplyr::mutate(
      phenotype = phenotype %>% 
      str_replace_all("[:digit:]", "") %>% 
      str_remove("\\..*"),
      n_individ = cases + controls
    ),
    by = c("phen1" = "phenotype")
  ) %>% 
  ggplot(
    aes(
      x = n_bivar, 
      y = n_individ,
      colour = fct_relevel(
        phen1,
        fct_disease
      )
    )
    ) +
  geom_point() +
  scale_colour_brewer(
    palette = "BrBG",
    type = "div",
    name = "GWAS"
    ) +
  labs(
    x = "Number of significant local genetic correlations",
    y = "Sample size"
    ) +
  theme_rhr

```

```{r chord-diagram-rho, fig.height = 6, fig.cap = "Chord diagram showing the number of significant bivariate local genetic correlations (p < 0.05/n_tests) between each of the phenotypes across all LD blocks. Positive and negative correlations are coloured red and blue, respectively."}

plot_bivar_chord_diagram(
    bivar_corr = bivar_bonf,
    fct_phen = fct_disease,
    palette = RColorBrewer::brewer.pal(n = 6, name = "BrBG")
  )


```

```{r asap-cosa-plot, eval = F, echo = F}

png(
  file = here::here("tmp", "20211012_trait_circlise.png"),
  bg = "transparent", 
  width = 150, 
  height = 150, 
  units = "mm",
  res = 300
)

plot_bivar_chord_diagram(
    bivar_corr = bivar_bonf,
    fct_phen = fct_disease,
    palette = RColorBrewer::brewer.pal(n = 6, name = "BrBG")
  )

dev.off()

```

```{r bivar-correlations-bonf-phen-count, fig.cap = "Number of significant local genetic correlations between trait pairs. The fill of the bars indicates the number of significant LD blocks for which the 95% confidence interval included 1."}

n_upper_r2_ci <-
  bivar_bonf %>% 
  dplyr::filter(
    r2.upper == 1
  ) %>% 
  dplyr::count(phen1, phen2, name = "n_upper_r2_ci") 

phen_count <-
  bivar_bonf %>% 
  dplyr::count(phen1, phen2, name = "total_n") %>% 
  dplyr::left_join(n_upper_r2_ci) %>% 
  dplyr::mutate(
    label = str_c(phen1, " vs ", phen2),
    n_upper_r2_ci =
      case_when(
        is.na(n_upper_r2_ci) ~ 0,
        TRUE ~ as.numeric(n_upper_r2_ci)
      ),
    n_upper_r2_ci_not_one =
      total_n - n_upper_r2_ci
  ) %>% 
  tidyr::pivot_longer(
    cols = contains("n_upper_r2")
  ) %>% 
  dplyr::mutate(
    name =
      case_when(
        name == "n_upper_r2_ci" ~ "TRUE",
        TRUE ~ "FALSE"
      )
  )
  
phen_count %>% 
  ggplot(
    aes(
      x = 
        fct_reorder(
          .f = label,
          .x = total_n,
          .fun = max,
          .desc = T
        ),
      y = value,
      fill = fct_rev(name)
    )
  ) +
  geom_col(
    colour = "black"
  ) +
  scale_fill_grey() +
  labs(
    x = "",
    y = "Number of significant\nlocal genetic correlations",
    fill = "Does upper limit of r2 95% confidence\n interval include 1?"
  ) +
  theme_rhr

```

## Signficant phenotype correlations at LD block level {.tabset}

### Text
- The LD block with the highest number of significant bivariate local $r_{g}$'s, LD block 758 and 1719, were positioned on chromosome 4 and 11, respectively (Figure \@ref(fig:corr-plot), \@ref(fig:edge-diagram)). 
- The locus on chromosome 4 contains three protein-coding genes (*GPM6A*, *WDR17* and *SPATA4*), while the LD block on chromosome 11 contains ten protein-coding genes, one of which includes *DRD2*, which encodes dopamine receptor D2. As perhaps expected, all three neuropsychiatric disorders positively correlate with one another at LD block 1719. Somewhat unexpected is the significant negative correlation between LBD and SCZ at LD block 1719.

```{r bivar-correlations-locus-count}

print("Top LD block when Bonferroni-correcting (p < 0.05/n_bivar_tests):")

bivar_bonf %>%
  dplyr::count(locus, chr, start, stop, n_snps) %>% 
  dplyr::arrange(-n) %>% 
  dplyr::inner_join(loci %>% 
                      janitor::clean_names() %>% 
                      dplyr::rename(
                        locus = loc
                      )) %>% 
  dplyr::slice_max(order_by = n, n = 1)

```

### Figures

```{r corr-plot, fig.height=16, fig.cap = "Heatmaps showing the standardised coefficient for genetic correlation (rho) for each association within the LD block with significant bivariate local genetic correlations (p < 0.05/n_tests). Significant negative and positive correlations are indicated by blue and red fill, respectively. Non-significant correlations have a grey fill."}

results$bivar %>% 
  dplyr::filter(locus %in% unique(bivar_bonf$locus)) %>% 
  dplyr::mutate(
    rho_fill = 
      case_when(
             p < bivar_thres ~ round(rho, 2)
           )
  ) %>% 
  ggplot(
    aes(
      x = phen1,
      y = phen2,
      fill = rho_fill,
      label = round(rho, 2)
    )
  ) +
  geom_tile(colour = "black") +
  geom_text(
    size = 2
    ) +
  facet_wrap(vars(locus), ncol = 5) +
  scale_fill_distiller(palette = "RdBu", direction = -1, na.value = "#cccccc", limits = c(-1, 1)) +
  theme_rhr
  
```

```{r, edge-diagram, fig.height = 14, fig.cap = "Edge diagrams for each LD block showing the standardised coefficient for genetic correlation (rho) for each significant bivariate local genetic correlations (p < 0.05/n_tests). Significant negative and positive correlations are indicated by blue and red colour, respectively."}

plot_edge_diagram(
  bivar_corr = 
    bivar_bonf,
  phen = fct_disease, 
  ncol = 3,
)

```

```{r asap-subgroup-plot, echo = F, eval = F}

data_to_plot <- 
  bivar_bonf %>% 
  dplyr::filter(
    locus %in% c(681,1273,1719,2351)
    # locus %in% c(681,2351)
  ) %>% 
  dplyr::mutate(
    locus_labels =
      stringr::str_c(
        "chr", chr, ": ", scales::comma(start), "-", scales::comma(stop)
      )
  )

locus_labels <- 
  setNames(
    unique(data_to_plot$locus_labels),
    nm = unique(data_to_plot$locus)
  )

plot <- 
  plot_edge_diagram(
    bivar_corr = 
      data_to_plot,
    phen = fct_disease, 
    locus_labels = locus_labels,
    multiple_corr = FALSE,
    ncol = 4,
    seed = 78
  ) + 
  theme(
    legend.position = "right"
  )

ggsave(
  plot,
  path = here::here("tmp"), 
  filename = "20220207_edgediagrams.png",
  device = "png", 
  width = 250, 
  height = 70, 
  dpi = 300, 
  units = "mm"
)

```


```{r individual-locus-plots, message = F, fig.height = 8}

loci_gr <-
  bivar_bonf %>%
  dplyr::filter(
    phen1 %in% c("AD", "LBD", "PD"),
    phen2 %in% c("AD", "LBD", "PD")
  ) %>% 
  dplyr::count(locus, chr, start, stop, n_snps) %>%
  dplyr::arrange(locus) %>% 
  GenomicRanges::makeGRangesFromDataFrame(
    .,
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    seqinfo = NULL,
    seqnames.field = "chr",
    start.field = "start",
    end.field = "stop"
  )

fig_list = vector(mode = "list", length = length(loci_gr))

for(i in 1:length(loci_gr)){
  
  fig_list[[i]] <- 
    plot_locus(
    locus_gr = loci_gr[i], 
    ref = ref
    )
  
  names(fig_list)[i] <- str_c("locus_", loci_gr[i]$locus)
  
}

fig_list


```

## LD blocks with known neurodegenerative pleiotropic genes {.tabset}

### Text
- Pleiotropic genes include:
    - APOE (implicated in AD and LBD)
    - TMEM175 (implicated in PD, LBD, and RBD)
    - SNCA (implicated in PD, LBD and RBD)
    - BIN1 (implicated in AD and LBD)
    - GRN (implicated in AD, PD, FTD, and LBD)
    - KIF5A (implicated in PD and ALS/FTD)
- AD and PD had sufficient local genetic signal (i.e. univariate heritability) at all tested LD blocks. LBD had insufficient local heritability in the GRN-, KIF5A and TMEM175-containing LD blocks, and therefore was not carried forward in these loci for bivariate testing.
- Of the LD blocks that contained implicated genes, only the SNCA- and APOE-containing LD blocks contained local $r_{g}$s that passed the Bonferroni significance cut-off (p < 0.05/n_tests; Figure \@ref(fig:pleiotropic-genes-fig)a). Using a more lenient FDR correction, local $r_{g}$s were observed across all LD blocks, except the GRN- and TMEM175-containing LD blocks (Figure \@ref(fig:pleiotropic-genes-fig)b).
    - BIN1-containing LD block 319
        - Bonferroni: no local $r_{g}$s. 
        - FDR: positive local $r_{g}$ between AD and LBD.
    - SNCA-containing LD block 681
        - Bonferroni & FDR: negative local $r_{g}$ between AD and PD.
    - KIF5A-containing LD block 1794
        - Bonferroni: no local $r_{g}$s.  
        - FDR: positive local $r_{g}$s between BIP and SCZ.
    - APOE-containing LD block 2351
        - Bonferroni: positive local $r_{g}$ observed between AD and LBD and negative local $r_{g}$ between PD and LBD.
        - FDR: as above and a negative local $r_{g}$ betyween AD and PD.
- In terms of expectations, APOE and BIN1 are what we would expect i.e. positive local $r_{g}$ between AD and LBD, but some unexpected local $r_{g}$s too e.g. negative local $r_{g}$ between AD and PD in APOE- and SNCA-containing LD blocks.

### Tables
```{r pleiotropic-genes-table}

print("LD blocks containing pleiotropic genes:")

loci_genes_filtered %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

print("Univariate results filtered for neurodegenerative phenotypes AND LD blocks containing pleiotropic genes. Significant when p < 0.05/n_ld_blocks.")

results$univ %>% 
  dplyr::inner_join(
    loci_genes_filtered %>% 
      dplyr::select(locus, gene_name),
    by = c("locus")
  ) %>% 
  dplyr::filter(
    phen %in% c("AD", "LBD", "PD")
  ) %>% 
  dplyr::mutate(
    significant =
      case_when(
        p < 0.05/nrow(loci) ~ TRUE,
        TRUE ~ FALSE
      )
  ) %>% 
  dplyr::select(
    locus, gene_name, significant, everything()
  ) %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')

print("Bivariate results filtered for neurodegenerative phenotypes AND LD blocks containing pleiotropic genes.")

results$bivar %>% 
  dplyr::inner_join(
    loci_genes_filtered %>% 
      dplyr::select(locus, gene_name),
    by = c("locus")
  ) %>% 
  dplyr::filter(
    phen1 %in% c("AD", "LBD", "PD") & phen2 %in% c("AD", "LBD", "PD")
  ) %>% 
  dplyr::mutate(
    bonferroni =
      case_when(
        p < bivar_thres ~ TRUE,
        TRUE ~ FALSE
      ),
    fdr =
      case_when(
        fdr < 0.05 ~ TRUE,
        TRUE ~ FALSE
      )
  ) %>% 
  dplyr::select(
    locus, gene_name, bonferroni, fdr, everything()
  ) %>%
  DT::datatable(rownames = FALSE,
                options = list(scrollX = TRUE),
                class = 'white-space: nowrap')


```



### Figures
```{r pleiotropic-genes-fig, fig.height = 8, fig.cap = "Edge diagrams for each LD block showing local genetic correlations when corrected using (a) the stringent Bonferroni  correction or (b) the lenient FDR correction. The standardised coefficient for genetic correlation (rho) is indicated for significant local correlations. Negative and positive correlations are indicated by blue and red colour, respectively."}
a <- 
  plot_edge_diagram(
    bivar_corr = 
      results$bivar %>% 
      dplyr::inner_join(
        loci_genes_filtered %>% 
          dplyr::select(locus, gene_name),
        by = c("locus")
      ) %>% 
      dplyr::mutate(
        locus = str_c(locus, ": contains ", gene_name)
      ),
    p_threshold = bivar_thres,
    multiple_corr = FALSE,
    phen = fct_disease, 
    ncol = 3
  )

b <- 
  plot_edge_diagram(
    bivar_corr = 
      results$bivar %>% 
      dplyr::inner_join(
        loci_genes_filtered %>% 
          dplyr::select(locus, gene_name),
        by = c("locus")
      ) %>% 
      dplyr::mutate(
        locus = str_c(locus, ": contains ", gene_name)
      ) %>% 
      dplyr::filter(fdr < 0.05),
    multiple_corr = FALSE,
    phen = fct_disease, 
    ncol = 3
  ) +
  theme(legend.position = "none")

cowplot::plot_grid(
  a,b,
  ncol = 1, 
  rel_heights = c(1,2), 
  align = "v", 
  axis = "lr",
  labels = letters[1:2]
)
```



<br><br>

# Session info

<details>
  <summary>Show/hide</summary>

```{r reproducibility, echo = FALSE}
# Session info
library("sessioninfo")
options(width = 120)
session_info()
```

</details> 
