---
title: "Manuscript figures"
author: 
- name: "Regina H. Reynolds"
  affiliation: UCL
output: 
  bookdown::html_document2:
    figure_caption: yes
    code_folding: hide
    theme: paper
    highlight: kate
    df_print: paged
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include = FALSE}

library(circlize) # For chord diagram
library(cowplot) # For plot arrangement
library(ggplot2) # For plotting
library(ggraph) # For edge diagrams
library(tidyverse) # For tidy manipulation of data
library(stringr) # For string manipulation
library(rtracklayer) # For loading of reference gtf

knitr::opts_chunk$set(echo = T, warning = F, message = F)

# Set defaults for ggplots 
source(here::here("R", "theme_rhr.R"))

loci <- LAVA::read.loci(here::here("results", "01_input_prep", "gwas_filtered.loci"))
fct_disease <- c("AD", "LBD", "PD", "BIP", "MDD", "SCZ")

```

> Aim: generate figures for draft manuscript
<br><br>

# Supplementary code {.tabset}
Following section includes any intermediary code used in this `.Rmd`.

## Create output directory
```{r outdir}

out_dir <- here::here("results", "99_manuscript_figures")
dir.create(out_dir, showWarnings = T)

```

## Loading plotting functions

```{r load-plot-func}

source(here::here("R", "plot_global_corr.R"))
source(here::here("R", "plot_chord_diagram.R"))
source(here::here("R", "plot_locus.R"))
source(here::here("R", "plot_edge_diagram.R"))

```

# Figure 1 {.tabset}

Overview of  local and global genetic correlations between neurodegenerative diseases and neuropsychiatric disorders. 

## Load data

```{r fig1-data}

# LAVA results
results <- readRDS(here::here("results", "02_univar_bivar_test", "results_summary.rds"))

# LDSC results
global <- 
  read_delim(
    file = here::here("results", "01_input_prep", "ldsc_corr", "ldsc_correlations.txt"),
    delim = "\t"
    ) 

# GWAS loci
gwas_loci <- 
  readRDS(here::here("results", "01_input_prep", "gwas_filtered_loci.rds"))

# LD blocks
ld_blocks <- 
  readRDS(here::here("results", "01_input_prep", "gwas_filtered_loci_w_genes.rds")) %>% 
  dplyr::ungroup()

```

## Data wrangling

```{r fig1-wrangle}

p_threshold <- 0.05/nrow(results$bivar)

# Tidy labels
bivar_tidy <- 
  results$bivar %>% 
  dplyr::inner_join(
    ld_blocks %>% 
      dplyr::select(locus, assoc_gwas)
  ) 

global_tidy <- 
  global %>%
  dplyr::mutate(
    p1 = p1 %>%
      stringr::str_replace_all("[:digit:]", "") %>%
      stringr::str_remove("\\..*"),
    p2 = p2 %>%
      stringr::str_replace_all("[:digit:]", "") %>%
      stringr::str_remove("\\..*")
  )

# N of phen with CI 97.5 == 1
n_upper_r2_ci <-
  bivar_tidy %>% 
  dplyr::filter(
    p < p_threshold,
    r2.upper == 1
  ) %>% 
  dplyr::count(phen1, phen2, name = "n_upper_r2_ci") 

bivar_count <-
  bivar_tidy %>% 
  dplyr::filter(
    p < p_threshold
  ) %>% 
  dplyr::count(phen1, phen2, name = "total_n") %>% 
  dplyr::left_join(n_upper_r2_ci) %>% 
  dplyr::mutate(
    label = str_c(phen1, " vs ", phen2),
    n_upper_r2_ci =
      case_when(
        is.na(n_upper_r2_ci) ~ 0,
        TRUE ~ as.numeric(n_upper_r2_ci)
      ),
    n_upper_r2_ci_not_one =
      total_n - n_upper_r2_ci
  ) %>% 
  tidyr::pivot_longer(
    cols = contains("n_upper_r2")
  ) %>% 
  dplyr::mutate(
    name =
      case_when(
        name == "n_upper_r2_ci" ~ "TRUE",
        TRUE ~ "FALSE"
      )
  )

```

## Main figure {.active}

```{r fig1, fig.height = 6}

plot <- vector(mode = "list", length = 2)

plot[[1]] <- 
  plot_global_bivar(
    global_corr = 
      global_tidy,
    bivar_corr = 
      bivar_tidy,
    n_phenotypes = 6
  ) + 
  theme(legend.position = "bottom")

plot[[2]] <- 
  bivar_count %>% 
  ggplot(
    aes(
      x = 
        fct_reorder(
          .f = label,
          .x = total_n,
          .fun = max,
          .desc = T
        ),
      y = value,
      fill = name
    )
  ) +
  geom_col(
    colour = "black"
  ) +
  scale_fill_grey(breaks=c("TRUE", "FALSE")) +
  labs(
    x = "",
    y = "Number of significant\nlocal genetic correlations",
    fill = "Does upper limit of\nr2 95% CI include 1?"
  ) +
  theme_rhr +
  theme(legend.position = "bottom")

fig <-
  cowplot::plot_grid(
    cowplot::plot_grid(
      NULL, plot[[1]],
      labels = letters[1:2]
    ),
    plot[[2]],
    ncol = 1,
    labels = c("", "c"),
    rel_heights = c(1.5,1)
  )
  
  

fig

```

```{r fig1-save, eval = F}

png(
  file = file.path(out_dir, "fig_1a_global_local.png"),
  bg = "transparent", 
  width = 120, 
  height = 120, 
  units = "mm",
  res = 300
)

plot_bivar_chord_diagram(
    bivar_corr = 
      bivar_tidy %>% 
      dplyr::filter(
        p < p_threshold
      ),
    fct_phen = fct_disease,
    palette = RColorBrewer::brewer.pal(n = 6, name = "BrBG")
  )

dev.off()

ggsave(
  fig,
  path = out_dir, 
  filename = "fig_1bc_global_local.png",
  device = "png", 
  width = 150, 
  height = 180, 
  dpi = 300, 
  units = "mm"
)

```

## Supplementary figures

```{r fig1-supp-fig}

plot <- list()

plot[[1]] <- 
  gwas_loci %>% 
  dplyr::distinct(LD_LOC, LD_CHR, LD_start, LD_end) %>% 
  dplyr::count(LD_CHR) %>% 
  ggplot(
    aes(
      x = 
        forcats::fct_reorder(
          .f = as.factor(LD_CHR), 
          .x = n, 
          .fun = median, 
          .desc = FALSE
        ),
      y = n
    )
  ) +
  geom_col(colour = "black", fill = "#d9d9d9") +
  labs(
    x = "Chromosome",
    y = "Number of LD blocks"
  ) +
  coord_flip() + 
  theme_rhr + 
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))

plot[[2]] <- 
  gwas_loci %>% 
  dplyr::count(LD_CHR, GWAS) %>% 
  dplyr::mutate(
    LD_CHR = as.factor(LD_CHR),
    GWAS = str_to_upper(GWAS)
  ) %>% 
  ggplot(
    aes(
      x = fct_rev(LD_CHR),
      y = n,
      fill = GWAS
    )
  ) +
  geom_col(
    position = position_dodge2(preserve = "single"),
    colour = "black"
  ) +
  scale_fill_brewer(
    palette = "BrBG",
    type = "div"
  ) +
  labs(
    x = "Chromosome",
    y = "Number of SNPs"
  ) + 
  coord_flip() + 
  theme_rhr +
  theme(
    axis.text.x = element_text(angle = 0, hjust = 0.5),
    legend.position = "right"
    )

suppfig <- 
  cowplot::plot_grid(
    plotlist = plot,
    align = "v",
    axis = "lr",
    ncol = 1,
    nrow = 2,
    rel_heights = c(1,2),
    labels = letters[1:length(plot)]
  )

suppfig

```

```{r fig1-supp-fig-save, eval = F}

ggsave(
  suppfig,
  path = out_dir, 
  filename = "suppfig_1_ldblocks.png",
  device = "png", 
  width = 150, 
  height = 180, 
  dpi = 300, 
  units = "mm"
)

```


## Supplementary tables

### LD blocks

```{r fig1-supp-tables-ldblocks, eval = F}

xlsx <- 
  setNames(
    vector(mode = "list", length = 2),
    c("column_descriptions", "results")
  )

xlsx[[2]] <-
  ld_blocks
  
xlsx[[1]] <-
  tibble(
    column_name = colnames(ld_blocks),
    description = 
      c(
        "LD block ID",
        "LD block chromosome",
        "LD block start base pair",
        "LD block end base pair",
        "Number of genes overlapping LD block",
        "Genes overlapping LD block (ensembl ID)",
        "Genes overlapping LD block (gene name)",
        "Number of GWAS with genome-wide significant SNPs overlapping LD block",
        "GWAS with genome-wide significant SNPs overlapping LD block"
      )
  )

openxlsx::write.xlsx(
  xlsx,
  file = file.path(out_dir, "SupplementaryTable1_ld_blocks.xlsx"),
  row.names = FALSE,
  headerStyle = openxlsx::createStyle(textDecoration = "BOLD"),
  firstRow = TRUE,
  append = TRUE,
  overwrite = TRUE
)

```


### Global correlations

```{r fig1-supp-tables-global, eval = F}

xlsx <- 
  setNames(
    vector(mode = "list", length = 2),
    c("column_descriptions", "results")
  )

xlsx[[2]] <-
  global

xlsx[[1]] <-
tibble(
  column_name = colnames(global),
  description = 
    c(
      "Phenotype 1",
      "Phenotype 2",
      "The estimated genetic correlation",
      "The bootstrap standard error of the genetic correlation estimate",
      "The bootstrap standard error of the genetic correlation estimate",
      "P-value for genetic correlation",
      "Estimated snp-heritability of the second phenotype ",
      "Standard error of h2 for phenotype 2",
      "Single-trait LD score regression intercept for phenotype 2",
      "Standard error for single-trait LD score regression intercept for phenotype 2",
      "Estimated genetic covariance between p1 and p2",
      "Bootstrap standard error of cross-trait LD score regression intercept"
    )
)

openxlsx::write.xlsx(
  xlsx,
  file = file.path(out_dir, "SupplementaryTable2_ldsc_global.xlsx"),
  row.names = FALSE,
  headerStyle = openxlsx::createStyle(textDecoration = "BOLD"),
  firstRow = TRUE,
  append = TRUE,
  overwrite = TRUE
)

```

### Univariate/bivariate correlations

```{r fig1-supp-tables-lava, eval = F}

xlsx <- 
  setNames(
    vector(mode = "list", length = 3),
    c("column_descriptions", "univariate", "bivariate")
  )

xlsx[[2]] <-
  results$univ

xlsx[[3]] <-
  results$bivar %>% 
  dplyr::select(-fdr)

xlsx[[1]] <-
  tibble(
    column = colnames(results$univ),
    description = 
      c(
        "LD block ID",
        "LD block chromosome",
        "LD block start base pair",
        "LD block end base pair",
        "The number of SNPs within the LD block",
        "The number of PCs retained within the LD block",
        "Analysed phenotype",
        "Observed local heritability",
        "P-value from the univariate test (F-test for continuous, Chi-sq for binary)"
      )
  ) %>%
  dplyr::mutate(sheet = "univariate") %>% 
  dplyr::bind_rows(
    tibble(
      column = colnames(results$bivar),
      description = 
        c(
          "LD block ID",
          "LD block chromosome",
          "LD block start base pair",
          "LD block end base pair",
          "The number of SNPs within the LD block",
          "The number of PCs retained within the LD block",
          "Phenotype 1",
          "Phenotype 2",
          "Standardised coefficient for the local genetic correlation",
          "Lower 95% confidence estimate for rho",
          "Upper 95% confidence estimate for rho",
          "Equivalent of taking the square of rho. Denotes the proportion of variance in genetic signal for phen1 explained by phen2 (and vice versa)",
          "Lower 95% confidence estimate for r2",
          "Upper 95% confidence estimate for r2",
          "Simulation p-values for the local genetic correlation"
        )
    ) %>% 
      dplyr::mutate(sheet = "bivariate")
  ) %>% 
  dplyr::select(sheet, everything())

openxlsx::write.xlsx(
  xlsx,
  file = file.path(out_dir, "SupplementaryTable3_lava_local.xlsx"),
  row.names = FALSE,
  headerStyle = openxlsx::createStyle(textDecoration = "BOLD"),
  firstRow = TRUE,
  append = TRUE,
  overwrite = TRUE
)

```

# Figure 2 {.tabset}

## Data wrangling

```{r fig2-wrangle}

bivar_pair_type <-
  bivar_tidy  %>% 
  dplyr::filter(
    p < p_threshold
  ) %>%  
  dplyr::group_by(locus, phen1, phen2) %>% 
  dplyr::mutate(
    phen1_overlapping_assoc_gwas =
      any(phen1 %in% unlist(assoc_gwas)),
    phen2_overlapping_assoc_gwas =
      any(phen2 %in% unlist(assoc_gwas)),
    sum_phen_overlapping_assoc_gwas =
      sum(phen1_overlapping_assoc_gwas, phen2_overlapping_assoc_gwas),
    phen1_is_neurodegen = 
      phen1 %in% c("AD", "PD", "LBD"),
    phen2_is_neurodegen = 
      phen2 %in% c("AD", "PD", "LBD"),
    sum_phen_is_neurodegen =
      sum(phen1_is_neurodegen, phen2_is_neurodegen),
    trait_pair_type =
      case_when(
        sum_phen_is_neurodegen == 2 ~ "Both neurodegenerative",
        sum_phen_is_neurodegen == 1 ~ "Mixed",
        sum_phen_is_neurodegen == 0 ~ "Both neuropsychiatric"
      )
  ) %>% 
  dplyr::group_by(locus, phen1, phen2) %>% 
  dplyr::mutate(
    locus_labels =
      stringr::str_c(
        "LD block ", locus, 
        " (", str_c(unlist(assoc_gwas), collapse = ", "), ")",
        "\nchr", chr, ": ", scales::comma(start), "-", scales::comma(stop)
      )
  )

locus_labels <- 
  setNames(
    nm = unique(bivar_pair_type$locus),
    object = unique(bivar_pair_type$locus_labels)
  )

```

## Main figure {.active}

```{r fig2}

plot <- vector(mode = "list", length = 4)

plot[[1]] <- 
  bivar_pair_type %>% 
  ggplot(
    aes(
      x = as.factor(sum_phen_overlapping_assoc_gwas),
      fill = trait_pair_type
    )
  ) + 
  geom_bar(colour = "black") +
  scale_fill_manual(
    values = RColorBrewer::brewer.pal(n = 3, name = "BrBG")[c(1,3,2)],
    name = "Traits in pair"
  ) + 
  labs(x = "Number of traits in pair with genome-\nwide significant SNPs overlapping LD block", y = "Count") +
  coord_flip() +
  guides(
    fill = guide_legend(
      nrow = 3,
      title.position = "top"
      )
    ) +
  theme_rhr +
  theme(
    legend.position = "bottom",
    axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0)
  )

plot[[2]] <-
  ggplot +
  geom_blank()

plot[[3]] <-
  plot_edge_diagram(
    bivar_corr = 
      bivar_pair_type %>% 
      dplyr::filter(locus %in% c(1719, 2281)),
    phen = fct_disease, 
    locus_labels = locus_labels,
    multiple_corr = FALSE,
    geom_node_size = 2,
    geom_label_size = 1.7,
    base_size = 8,
    ncol = 1,
    seed = 57 #26,35
  ) +
  theme(
    legend.position = "none"
  )

plot[[4]] <-
  bivar_tidy %>% 
  dplyr::filter(
    locus %in% c(681, 1273, 2351),
    phen1 %in% c("AD", "LBD", "PD"),
    phen2 %in% c("AD", "LBD", "PD")
  ) %>% 
  dplyr::mutate(
    rho_fill = 
      case_when(
        p < p_threshold ~ round(rho, 2)
      ),
    rho_label =
      case_when(
        locus == 1273 & phen1 == "AD" & phen2 == "PD" |
          locus == 2351 & phen1 == "AD" & phen2 == "LBD" ~ str_c(round(rho, 2), "*"),
        TRUE ~ as.character(round(rho, 2))
      )
  ) %>% 
  ggplot(
    aes(
      x = phen1,
      y = phen2
    )
  ) +
  geom_tile(
    aes(fill = rho_fill),
    colour = "black"
  ) +
  geom_text(
    aes(label = rho_label),
    size = 2.5
  ) +
  labs(x = "", y = "", fill = "rho") +
  facet_wrap(
    vars(locus), 
    ncol = 3,
    labeller = as_labeller(locus_labels)
  ) +
  scale_fill_distiller(
    palette = "RdYlBu", 
    direction = -1, 
    na.value = "#cccccc", 
    limits = c(-1, 1)
    ) +
  theme_rhr + 
  theme(
    axis.text.x = element_text(angle = 0, hjust = 0.5),
    panel.grid = element_blank()
    )

fig <-
  cowplot::plot_grid(
    cowplot::plot_grid(
      plotlist = plot[c(1:3)],
      ncol = 3,
      nrow = 1,
      # align = "hv",
      # axis = "bt",
      rel_widths = c(1,0.8,1.5),
      labels = c("", "b", ""), 
      # move "b" to right of middle blank plot
      label_x = 0.9, label_y = 1 
    ),
    plot[[4]],
    ncol = 1,
    nrow = 2,
    # align = "v",
    # axis = "lr",
    rel_heights = c(1.5,1),
    labels = c("a", "c")
  )


fig
```

```{r fig2-save, eval = F}

ggsave(
  fig,
  path = out_dir, 
  filename = "fig_2_local_regions.png",
  device = "png", 
  width = 180, 
  height = 160, 
  dpi = 300, 
  units = "mm"
)

```


## Supplementary figure(s)

Refer to [05_run_noproxy_check.html](https://rhreynolds.github.io/neurodegen-psych-local-corr/05_run_noproxy_check.html).

## Supplementary table(s)

Refer to [05_run_noproxy_check.html](https://rhreynolds.github.io/neurodegen-psych-local-corr/05_run_noproxy_check.html).

# Figure 3 {.tabset}

## Load data

```{r fig3-data}

multireg <-
  list.files(
    here::here(
      "results", 
      "03_partial_corr_multi_reg"
      ),
    pattern = ".multireg.lava.rds", 
    full.names = T
  ) %>% 
  readRDS() %>% 
  lapply(., function(x){
    
    x %>% 
      lapply(., function(y){
        
        y %>% 
          lapply(., function(z){
            
            z %>% 
              qdapTools::list_df2df(col1 = "model_number")
            
          }) %>% 
          qdapTools::list_df2df(col1 = "list_name")
        
      }) %>% 
      qdapTools::list_df2df()
    
  }) %>% 
    qdapTools::list_df2df(col1 = "locus") %>% 
    dplyr::select(-X1) %>% 
    dplyr::mutate(
      model_type = 
        case_when(
          locus == 1719 & list_name == "L1" ~ "intermediate",
          locus == 1719 & list_name == "L2" ~ "full",
          TRUE ~ "full"
          )
    ) %>% 
    dplyr::select(
      locus, contains("model"), 
      outcome, predictors, 
      everything(), -contains("list")
      )

```

## Data wrangling

```{r fig3-wrangle}

multireg_tidy <-
  # filter outcomes w. significant predictors
  multireg %>% 
  dplyr::filter(
    p < 0.05
  ) %>%
  dplyr::select(locus, contains("model"), outcome) %>% 
  dplyr::distinct() %>% 
  # join all predictors
  dplyr::inner_join(
    multireg,
    c("locus", "model_number", "model_type", "outcome")
  ) %>% 
  dplyr::mutate(
    predictors = predictors %>%
      stringr::str_replace_all("[:digit:]", "") %>%
      stringr::str_remove("\\..*"),
    outcome = outcome %>% 
      stringr::str_replace_all("[:digit:]", "") %>%
      stringr::str_remove("\\..*"),
    p_text = case_when(
      p < 0.001 ~ "***",
      p < 0.01 ~ "**",
      p < 0.05 ~ "*"
    ),
    locus = 
      as.numeric(locus)
  )

multireg_tidy <- 
  multireg_tidy %>% 
  dplyr::inner_join(
    multireg_tidy %>% 
      dplyr::group_by(outcome, locus) %>% 
      dplyr::summarise(predictors = str_c(predictors, collapse = " + ")) %>% 
      dplyr::mutate(model = str_c(outcome, " ~ ", predictors)) %>% 
      dplyr::select(outcome, locus, model)
  )

```

## Main figure {.active}

```{r fig3}

plot <- vector(mode = "list", length = 2)

plot[[1]] <- 
  multireg_tidy %>% 
  ggplot(
    aes(
      x = predictors, 
      y = gamma, 
      ymin = gamma.lower, 
      ymax = gamma.upper
      )
    ) +
  geom_pointrange(
    colour = "#888888", fatten = 0.25
    ) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_text(
    aes(
      y = gamma.upper + (0.1 * gamma.upper), 
      label = p_text
      )
    ) +
  facet_wrap(
    vars(locus, model), 
    scales = "free_x",
    nrow = 1
    ) +
  labs(x = "Predictors", y = "Standardised multiple\nregression coefficient") +
  theme_rhr

plot[[2]] <- 
  multireg_tidy %>% 
  dplyr::mutate(
    locus_model =
      str_c(locus, model, sep = "\n")
  ) %>% 
  dplyr::distinct(
    locus, locus_model, r2, r2.lower, r2.upper
  ) %>% 
  ggplot(
    aes(
      x = locus_model %>% 
        fct_reorder(., .x = as.numeric(locus), .fun = max),
      y = r2,
      ymin = r2.lower,
      ymax = r2.upper
    )
  ) +
  geom_col(colour = "black", fill = "#d9d9d9") + 
  geom_errorbar(width=.2) +
  labs(
    x = "Locus and model",
    y = "Multivariate r2"
  ) +
  theme_rhr

  
fig <- 
  cowplot::plot_grid(
    plotlist = plot, 
    ncol = 1, 
    axis = "lr", 
    align = "v",
    labels = letters[1:length(plot)]
  )

fig
```

```{r fig3-save, eval = F}

ggsave(
  fig,
  path = out_dir, 
  filename = "fig_3_multireg.png",
  device = "png", 
  width = 130, 
  height = 120, 
  dpi = 300, 
  units = "mm"
)

```

## Supplementary tables

```{r fig3-supp-tables , eval = F}

xlsx <- 
  setNames(
    vector(mode = "list", length = 2),
    c("column_descriptions", "results")
  )

xlsx[[2]] <-
  multireg

xlsx[[1]] <-
  tibble(
    column = colnames(multireg),
    description = 
      c(
        "LD block ID",
        "For multiple regressions with multiple intermediate models, this is a model identifier",
        "Whether the regression model is an intermediate or full model",
        "Outcome phenotype",
        "Predictor phenotype",
        "Standardised multiple regression coefficient",
        "Lower bound of 95% confidence interval for gamma",
        "Upper bound of 95% confidence interval for gamma",
        "Proportion of variance in genetic signal for the outcome phenotype explained by all predictor phenotypes simultaneously",
        "Lower bound of 95% confidence interval for r2",
        "Upper bound of 95% confidence interval for r2",
        "Simulation p-values for the gammas"
      )
  )

openxlsx::write.xlsx(
  xlsx,
  file = file.path(out_dir, "SupplementaryTable5_multireg.xlsx"),
  row.names = FALSE,
  headerStyle = openxlsx::createStyle(textDecoration = "BOLD"),
  firstRow = TRUE,
  append = TRUE,
  overwrite = TRUE
)


```


# Figure 4 {.tabset}

## Load data

```{r fig4-data}

# Genic regions
gene_filtered_loci <- 
  readRDS(
    here::here("results", "04_eqtl_univar_bivar", "window_100000", "gene_filtered_loci.rds")
  )

# LAVA results
qtl_tidy <- readRDS(here::here("results", "04_eqtl_univar_bivar", "results_summary.rds"))

# Reference gene annotation
ref <- import("/data/references/ensembl/gtf_gff3/v87/Homo_sapiens.GRCh37.87.gtf")
ref <- ref %>% keepSeqlevels(c(1:22), pruning.mode = "coarse") 
ref <- ref[ref$type == "gene"]

```

## Data wrangling

```{r fig4-wrangle}

qtl_to_plot <- 
  qtl_tidy$bivar %>% 
  lapply(.,function(df){
    
    df %>% 
      dplyr::filter(
        str_detect(phen2, "ENSG")
      ) %>% 
      dplyr::mutate(
        fill_rho = case_when(
          p < 0.05 ~ round(rho, 2)
        ),
        sig = case_when(
          fdr < 0.05 ~ "**"
        ),
        nom_sig = case_when(
          p < 0.05 & fdr > 0.05 ~ ".",
        ),
        chr_gene = str_c("chr ",chr,": ", gene_name)
      )
    
  })

# Filter for genic regions containing significant eqtl-gwas local rg
# Include all significant local rgs in genic region 
qtl_tidy_filtered <-
  qtl_tidy$bivar$window_100000 %>% 
  dplyr::filter(fdr < 0.05, str_detect(phen2, "ENSG")) %>% 
  dplyr::distinct(ld_block, eqtl_dataset, gene_locus) %>% 
  dplyr::inner_join(qtl_tidy$bivar$window_100000) %>% 
  dplyr::filter(fdr < 0.05)

# Create list with results split by LD block
qtl_list <- 
  setNames(
    qtl_tidy_filtered %>% 
      dplyr::group_split(ld_block),
    nm = 
      str_c("locus_", unique(qtl_tidy_filtered$ld_block))
  )

# Number of genes implicated in each ld block
n_genes <- 
  gene_filtered_loci %>% 
  dplyr::ungroup() %>% 
  dplyr::select(ld_block = locus, gene_locus = gene_id) %>% 
  dplyr::distinct(ld_block, gene_locus) %>% 
  dplyr::count(ld_block, name = "n_genes") %>% 
  dplyr::inner_join(
    qtl_tidy_filtered %>% 
      dplyr::distinct(ld_block, gene_locus) %>% 
      dplyr::count(ld_block, name = "n_signif_genes")
  ) %>% 
  dplyr::mutate(n_not_signif_genes = n_genes - n_signif_genes) %>% 
  tidyr::pivot_longer(
    cols = contains("signif"),
    names_to = "signif",
    values_to = "n"
  ) %>% 
  dplyr::mutate(
    signif =
      case_when(
        signif == "n_signif_genes" ~ "TRUE",
        TRUE ~ "FALSE"
      )
  )

# Number of distinct eQTL-disease rgs per LD block 
# Use this to determine number of gene eQTLs shared across disease traits
n_shared_eqtl <- 
  qtl_tidy_filtered %>% 
  dplyr::filter(str_detect(phen2, "ENSG")) %>% 
  dplyr::distinct(ld_block, phen1, phen2) %>% 
  dplyr::count(ld_block, phen2, name = "n_disease") %>% 
  dplyr::count(ld_block, n_disease, name = "n_genes") %>% 
  dplyr::mutate(
    shared_eqtl =
      case_when(
        n_disease == 1 ~ "FALSE",
        n_disease > 1 ~ "TRUE"
      )
  )

# Pivot results
qtl_bivar_long <- 
  qtl_tidy$bivar %>% 
  qdapTools::list_df2df(col1 = "window_size") %>% 
  tidyr::pivot_longer( 
    cols = -c("window_size", "ld_block", "eqtl_dataset", "gene_locus", "gene_name", "phen1", "phen2"),
    names_to = "feature",
    values_to = "value"
  ) %>% 
  tidyr::pivot_wider(names_from = "window_size") %>% 
  dplyr::mutate(
    across(
      .cols = contains("window"),
      ~ case_when(
        feature == "p" ~ -log10(.x),
        TRUE ~ .x
      )
    )
  ) %>%
  dplyr::mutate(
    feature = 
      case_when(
        feature == "p" ~ "-log10(p)",
        TRUE ~ feature
      ), 
    phenotype_corr =
      case_when(
        str_detect(phen2, "ENSG") ~ "gwas-eqtl",
        TRUE ~ "gwas-gwas"
      ),
    doe_100000 =
      case_when(
        feature == "rho" & window_100000 < 0 ~ "-",
        feature == "rho" & window_100000 > 0 ~ "+"
      ),
    doe_50000 =
      case_when(
        feature == "rho" & window_50000 < 0 ~ "-",
        feature == "rho" & window_50000 > 0 ~ "+" 
      ),
    doe_same =
      case_when(
        doe_100000 == doe_50000 ~ TRUE,
        doe_100000 != doe_50000 ~ FALSE
      )
  )

# Create granges of ld blocks
loci_gr <-
  loci %>%
  dplyr::rename(locus = LOC) %>% 
  dplyr::filter(locus %in% qtl_tidy_filtered$ld_block) %>% 
  GenomicRanges::makeGRangesFromDataFrame(
    .,
    keep.extra.columns = TRUE,
    ignore.strand = TRUE,
    seqinfo = NULL,
    seqnames.field = "CHR",
    start.field = "START",
    end.field = "STOP"
  )

```


## Main figure {.active}

```{r fig4, fig.height = 10}

margin <- ggplot2::margin(t = 2, r = 0, b = 0, l = 5, unit = "mm")

plots <- vector(mode = "list")

plots[[1]] <-
  n_genes %>% 
  ggplot(
    aes(
      x = as.factor(ld_block),
      y = n,
      fill = signif
    )
  ) +
  geom_col(colour = "black") +
  geom_hline(yintercept = 1, linetype = "dashed") +
  labs(
    x = "LD block",
    y = "Number of eQTL genes",
    fill = "Significant eQTL-GWAS\nlocal rg?"
  ) +
  scale_fill_grey(breaks=c("TRUE", "FALSE")) +
  guides(
    fill = guide_legend(
      nrow = 1,
      title.position = "top"
    )
  ) +
  theme_rhr +
  theme(
    legend.position = "bottom",
    plot.margin = margin
  )

plots[[2]] <- 
  n_shared_eqtl %>% 
  ggplot(
    aes(
      x = as.factor(ld_block),
      y = n_genes,
      fill = shared_eqtl
    )
  ) +
  geom_col(
    colour = "black"
  ) +
  geom_hline(yintercept = 1, linetype = "dashed") +
  labs(
    x = "LD block",
    y = "Number of eQTL genes",
    fill = "eQTL shared across\nmultiple disease traits?"
  ) +
  scale_fill_grey(breaks=c("TRUE", "FALSE")) +
  guides(
    fill = guide_legend(
      nrow = 1,
      title.position = "top"
    )
  ) +
  theme_rhr  +
  theme(
    legend.position = "bottom",
    plot.margin = margin
  )

ld_blocks <- 
  c(1273, 1719)

for(i in 1:length(ld_blocks)){
  
  plots[[i+2]] <-
   qtl_to_plot$window_100000 %>% 
      dplyr::filter(ld_block == ld_blocks[i], fdr < 0.05) %>% 
      ggplot(
        aes(
          x = 
            fct_reorder(
              gene_name,
              start
            ),
          y = phen1
        )
      ) +
      geom_count(colour = "black", shape = 15, size = 8) +
      geom_count(aes(colour = fill_rho), shape = 15, size = 7) +
      geom_text(aes(label = fill_rho), size = 2) +
      coord_fixed() +
      facet_grid(rows = vars(eqtl_dataset)) +
      labs(x = "eQTL gene", y = "Disease trait") +
      scale_colour_distiller(
        palette = "RdYlBu",
        limits = c(-1, 1), breaks = c(-1,0,1), na.value = "grey") +
      theme_rhr +
      theme(
        plot.margin = ggplot2::margin(t = 0, r = 0, b = 0, l = 0, unit = "mm"),
        legend.position = "none"
      )
  
}

plots[[5]] <- 
  plot_qtl_edge_diagram(
    bivar_corr_qtl = 
      qtl_list[["locus_1273"]] %>% 
      dplyr::filter(gene_name %in% c("CLU", "SCARA5")),
    phen = fct_disease,
    margin = margin,
    seed = 89
  )

fig <-
  cowplot::plot_grid(
    cowplot::plot_grid(
      cowplot::plot_grid(
        cowplot::plot_grid(
          plotlist = plots[c(1:2)],
          nrow = 1,
          labels = c("a", "b")
        ),
        plots[[3]],
        rel_heights = c(1.5,1),
        # align = "v", 
        # axis = "lr",
        ncol = 1,
        labels = c("", "d")
      ),
      ncol = 2,
      plots[[4]],
      rel_widths = c(1.9,1),
      labels = c("", "c")
    ),
    ggpubr::ggarrange(
      plots[[5]]$`CLU:EQTLGEN`,
      plots[[5]]$`SCARA5:EQTLGEN`,
      common.legend = TRUE,
      legend = "bottom",
      labels = c("e","")
    ),
    ncol = 1,
    rel_heights = c(1.75,1)
  )

fig

```

```{r fig4-save}

ggsave(
  fig,
  path = out_dir, 
  filename = "fig_4_eqtl.png",
  device = "png", 
  width = 150, 
  height = 200, 
  dpi = 300, 
  units = "mm"
)

```

## Supplementary figure(s)

```{r}

supp_fig <- vector(mode = "list", length = 3)

ld_blocks <- unique(qtl_tidy$bivar$window_100000$ld_block)

plots <- 
  setNames(
    object = vector(mode = "list", length = length(ld_blocks)),
    nm = str_c("ld", ld_blocks)
  )

for(block in ld_blocks){
  
  plots[[str_c("ld",block)]] <- 
    qtl_to_plot$window_100000 %>% 
    dplyr::filter(ld_block == block) %>% 
    ggplot(
      aes(
        x = 
          fct_reorder(
            gene_name,
            start
          ),
        y = phen1
      )
    ) +
    geom_count(colour = "black", shape = 15, size = 5) +
    geom_count(aes(colour = fill_rho), shape = 15, size = 4) +
    coord_fixed() +
    geom_text(aes(label=sig), size = 3, vjust = 0.75) +
    geom_text(aes(label=nom_sig), size = 5, vjust = 0.1) +
    facet_grid(rows = vars(eqtl_dataset)) +
    labs(x = "eQTL gene", y = "Disease trait") +
    scale_colour_distiller(
      palette = "RdYlBu",
      limits = c(-1, 1), breaks = c(-1,0,1), na.value = "grey") +
    guides(colour = guide_colourbar(title = "Local rg")) +
    theme_rhr +
    theme(
      plot.margin = ggplot2::margin(t = 0, r = 0, b = 0, l = 0, unit = "mm")
    )
  
}

supp_fig[[1]] <-
  ggpubr::ggarrange(
    ggpubr::ggarrange(
      plotlist = 
        plots[c(1:2)]  %>% 
        lapply(., function(plot){
          
          plot +
            theme(legend.position = "none")
          
        }),
      align = "h",
      labels = letters[1:2]
    ),
    ggpubr::ggarrange(
      plotlist = 
        plots[c(3:4)]  %>% 
        lapply(., function(plot){
          
          plot +
            theme(legend.position = "none")
          
        }),
      align = "h",
      labels = letters[3:4]
    ),
    plots[[5]],
    ncol = 1,
    heights = c(1,1.75,1),
    common.legend = T,
    labels = c("", "", "e")
  )

```


```{r fig4-supp-window-size, fig.height = 10}

overlap <- readRDS(here::here("results", "04_eqtl_univar_bivar", "fdr_overlap.rds"))

plots <- vector(mode = "list", length = 2)

plots[[1]] <- 
  qtl_tidy$bivar %>% 
  qdapTools::list_df2df(col1 = "window_size") %>% 
  tidyr::pivot_longer( 
    cols = -c("window_size", "ld_block", "eqtl_dataset", "gene_locus", "gene_name", "phen1", "phen2"),
    names_to = "feature",
    values_to = "value"
  ) %>% 
  dplyr::mutate(
    window_size =
      window_size %>% 
      readr::parse_number() %>% 
      format(scientific = FALSE) %>% 
      as.factor()
  ) %>% 
  dplyr::filter(
    feature %in% c("fdr"),
    value < 0.05
  ) %>% 
  dplyr::count(window_size, feature, name = "n_total") %>% 
  dplyr::mutate(
    unique =
      case_when(
        feature == "fdr" ~ n_total - (overlap %>% dplyr::filter(fdr_100000 < 0.05 & fdr_50000 < 0.05) %>% nrow())
      ),
    shared =
      n_total-unique
  ) %>% 
  tidyr::pivot_longer(
    cols = unique:shared,
    names_to = "sharing",
    values_to = "n"
  ) %>% 
  ggplot(
    aes(
      x = window_size,
      y = n,
      fill = sharing
    )
  ) + 
  geom_col(colour = "black") +
  scale_fill_grey() +
  guides(
    fill = guide_legend(
      nrow = 2,
      title.position = "top"
    )
  ) +
  labs(
    x = "Window size (kb)",
    y = "Number of significant bivariate local rg",
    fill = "Sharing of local rgs"
  ) +
  theme_rhr +
  theme(legend.position = "bottom")
  
  
plots[[2]] <- 
  qtl_bivar_long %>% 
  dplyr::filter(
    !feature %in%
      c("chr", "start", "stop", "fdr", "bonferroni", "n_pcs", "n_snps", "r2"),
    !str_detect(feature, "upper"),
    !str_detect(feature, "lower")
  ) %>%
  ggplot(
    aes(
      x = window_50000,
      y = window_100000
    )
  ) + 
  geom_point(
    aes(
      fill = doe_same
    ),
    shape = 21,
    colour = "black",
    alpha = 0.4
  ) + 
  geom_smooth(
    method = "lm",
    formula = "y~x",
    level = 0.99,
    colour = "black",
    fill = "grey"
  ) +
  ggpubr::stat_cor(
    method = "pearson",
    cor.coef.name = "R",
    size = 3
  ) +
  geom_abline(
    intercept = 0,
    linetype = "dashed",
    colour = "#EE442F"
  ) +
  facet_wrap(
    vars(phenotype_corr, feature),
    scales = "free"
  ) +
  labs(
    x = "50-kb window",
    y = "100-kb window",
    fill = "Same direction of effect?"
  ) +
  scale_fill_discrete(na.translate = F) +
  theme_rhr +
  guides(
    fill = guide_legend(
      nrow = 2,
      title.position = "top"
    )
  ) +
  theme(
    axis.text.x = element_text(angle = 0, hjust = 0.5),
    legend.position = "bottom"
    )

supp_fig[[2]] <-
  cowplot::plot_grid(
    plotlist = plots, 
    nrow = 1, 
    axis = "bt", 
    align = "h",
    rel_widths = c(1,3),
    labels = letters[1:length(plots)]
  )

supp_fig[[2]]

```

```{r manuscript-supp-fig-r2}

supp_fig[[3]] <-
  qtl_tidy_filtered %>% 
  dplyr::mutate(
    phenotype_corr =
      case_when(
        str_detect(phen2, "ENSG") ~ "gwas-eqtl",
        TRUE ~ "gwas-gwas"
      )
  ) %>% 
  ggplot(
    aes(
      x = phenotype_corr,
      y = r2
      )
    ) +
  geom_boxplot(fill = "#d9d9d9", outlier.alpha = 0) +
  ggbeeswarm::geom_beeswarm(alpha = 0.5) +
  labs(
    x = "Trait type in pair",
    y = "Explained variance (r2)"
  ) +
  theme_rhr

supp_fig[[3]]

```

```{r fig4-supp-save, eval = F}

ggsave(
  supp_fig[[1]],
  path = out_dir, 
  filename = "suppfig_3_eqtl_all.png",
  device = "png", 
  width = 170, 
  height = 280, 
  dpi = 300, 
  units = "mm"
)

ggsave(
  supp_fig[[2]],
  path = out_dir, 
  filename = "suppfig_4_windowsize_analysis.png",
  device = "png", 
  width = 150, 
  height = 150, 
  dpi = 300, 
  units = "mm"
)

ggsave(
  supp_fig[[3]],
  path = out_dir, 
  filename = "suppfig_5_r2.png",
  device = "png", 
  width = 90, 
  height = 90, 
  dpi = 300, 
  units = "mm"
)

```


## Supplementary table(s)

```{r, fig4-supp-table, fig.height = 10, eval = F}

ld_blocks <- 
  qtl_tidy$bivar$window_100000 %>% .[["ld_block"]] %>% unique() %>% str_c("ld", .)

# Full results
results_by_block <- 
  qtl_tidy %>% 
  lapply(., function(list_of_df){
    
    tmp <-
      list_of_df %>% 
      lapply(., function(df){
        
        
        setNames(
          object =
            df %>%
            dplyr::group_split(ld_block),
          nm = 
            df[["ld_block"]] %>% 
            unique() %>% 
            sort() %>% 
            str_c("ld", .)
        )
        
      })
    
  })

col_descriptions <-
  tibble(
    column = colnames(results_by_block$univ$window_100000$ld681),
    description = 
      c(
        "LD block ID",
        "eQTL dataset from which gene eQTLs are derived",
        "Ensembl ID for gene for which eQTLs were tested in genic region",
        "HGNC symbol for gene for which eQTLs were tested in genic region",
        "Genic region chromosome",
        "Genic region start base pair",
        "Genic region end base pair",
        "The number of SNPs within the genic region",
        "The number of PCs retained within the genic region",
        "Analysed phenotype",
        "Observed local heritability",
        "P-value from the univariate test (F-test for continuous, Chi-sq for binary)"
      )
  ) %>% 
  dplyr::mutate(sheet = "univariate") %>% 
  dplyr::bind_rows(
    tibble(
      column = colnames(results_by_block$bivar$window_100000$ld681),
      description = 
        c(
          "LD block ID",
          "eQTL dataset from which gene eQTLs are derived",
          "Ensembl ID for gene for which eQTLs were tested in genic region",
          "HGNC symbol for gene for which eQTLs were tested in genic region",
          "Genic region chromosome",
          "Genic region start base pair",
          "Genic region end base pair",
          "The number of SNPs within the genic region",
          "The number of PCs retained within the genic region",
          "Phenotype 1",
          "Phenotype 2. Gene expression traits (i.e. gene eQTLs) are denoted by their ensembl gene id",
          "Standardised coefficient for the local genetic correlation",
          "Lower 95% confidence estimate for rho",
          "Upper 95% confidence estimate for rho",
          "Equivalent of taking the square of rho. Denotes the proportion of variance in genetic signal for phen1 explained by phen2 (and vice versa)",
          "Lower 95% confidence estimate for r2",
          "Upper 95% confidence estimate for r2",
          "Simulation p-values for the local genetic correlation",
          "FDR-corrected simulation p-values",
          "Bonferroni-corrected simulation p-values"
        )
    ) %>% 
      dplyr::mutate(sheet = "bivariate")
  ) %>% 
  dplyr::select(sheet, column, description)

for(window in names(qtl_tidy$bivar)){
  
  # Create ld block figures ----
  qtl_list <- 
    setNames(
      qtl_tidy$bivar[[window]] %>% 
        dplyr::filter(fdr < 0.05) %>% 
        dplyr::group_split(ld_block),
      nm = 
        str_c(ld_blocks)
    )
  
  qtl_genes <-
    qtl_tidy$bivar[[window]] %>%
    dplyr::filter(fdr < 0.05) %>% 
    dplyr::group_by(eqtl_dataset, gene_locus) %>%
    dplyr::filter(str_detect(phen2, "ENSG")) %>% 
    .[["gene_locus"]] %>% 
    unique()
  
  fig_list <- vector(mode = "list", length = length(ld_blocks))
  
  for(block in ld_blocks){
    
    locus_plot <- 
      plot_locus(
        locus_gr = loci_gr[loci_gr$locus == str_remove(block, "ld")], 
        ref = ref,
        highlight_gene = qtl_genes,
        highlight_gene_label = "QTL - local rg?",
        window = str_remove(window, "window_") %>% as.numeric()
      )
    
    edge_plot <- 
      plot_qtl_edge_diagram(
        bivar_corr_qtl = qtl_list[[block]],
        phen = fct_disease,
        seed = 89
      )
    
    if(length(edge_plot)/4 > 1){
      
      height <- c(1,2.5)
      
    } else{
      
      height <- c(1,0.75)
    }
    
    fig_list[[which(ld_blocks == block)]] <- 
      ggpubr::ggarrange(
        locus_plot,
        ggpubr::ggarrange(
          plotlist = edge_plot,
          ncol = 4,
          nrow = ceiling(length(edge_plot)/4), 
          common.legend = TRUE,
          legend = "top"
        ),
        ncol = 1,
        labels = letters[1:2],
        heights = height
      )
    
    names(fig_list)[which(ld_blocks == block)] <- block
  }
  
  # Save to workbook ----
  wb <- openxlsx::createWorkbook()
  
  # Write column descriptions ----
  openxlsx::addWorksheet(wb, sheetName = "column_descriptions")
  
  openxlsx::writeData(
    wb,
    sheet = "column_descriptions",
    x = col_descriptions,
    row.names = FALSE,
    headerStyle = openxlsx::createStyle(textDecoration = "BOLD")
  )
  
  openxlsx::freezePane(
    wb,
    sheet = "column_descriptions",
    firstRow = TRUE
  )
  
  for(block in ld_blocks){
    
    # Write results ----
    for(test in names(results_by_block)){
      
      sheetname <- str_c(block, "_", test)
      
      openxlsx::addWorksheet(wb, sheetName = sheetname)
      
      openxlsx::writeData(
        wb,
        sheet = sheetname,
        x = results_by_block[[test]][[window]][[block]],
        row.names = FALSE,
        headerStyle = openxlsx::createStyle(textDecoration = "BOLD")
      )
      
      openxlsx::freezePane(
        wb,
        sheet = sheetname,
        firstRow = TRUE
      )
      
    }
    
    # Add figure ----
    print(fig_list[[block]])
    
    if(block == "ld2351"){
      
      height <- 45
      
    } else{
      
      height <- 20
    }
    
    openxlsx::insertPlot(
      wb,
      sheet = str_c(block, "_bivar"),
      width = 20,
      height = height,
      units = "cm",
      startRow = 2,
      startCol = 22
    )
    
  }
  
  openxlsx::saveWorkbook(
    wb,
    file = file.path(out_dir, str_c("SupplementaryTable_lava_local_eqtl_", window,".xlsx")),
    overwrite = TRUE
  ) 
  
}

```


<br><br>

# Session info

<details>
  <summary>Show/hide</summary>

```{r reproducibility, echo = FALSE}
# Session info
library("sessioninfo")
options(width = 120)
session_info()
```

</details> 
